### PR TITLE
guides: Add challenges page

### PR DESCRIPTION
**Pull request by StefanJum**

### CODE REVIEW

File: content/guides/challenges.mdx [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+title: Challenges
```
Comment (razvand) [inline content/guides/challenges.mdx:2]:
```suggestion
title: Hackathon Challenges
```

File: content/guides/challenges.mdx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+description: "Description of hackathon challenges."
```
Comment (razvand) [inline content/guides/challenges.mdx:3]:
```suggestion
description: "Listing of hackathon challenges, together with awarded points"
```

File: content/guides/challenges.mdx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+---
+
+## Challenges
```
Comment (razvand) [inline content/guides/challenges.mdx:6]:
```suggestion
## Hacakthon Challenges
```

File: content/guides/challenges.mdx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+You've made it so far.
```
Comment (razvand) [inline content/guides/challenges.mdx:8]:
```suggestion
You've been part of a Unikraft hackathon and you've made it so far.
```

File: content/guides/challenges.mdx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+Now the time has come to show the Unikraft world what you are really made of.
+A true Dragon Slayer!
+
+See the items added in the [`Hackathons` project](https://github.com/orgs/unikraft/projects/29).
```
Comment (razvand) [inline content/guides/challenges.mdx:12]:
```suggestion
See the items added in the [`Hackathons` GitHub project](https://github.com/orgs/unikraft/projects/29).
```

File: content/guides/challenges.mdx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+title: Challenges
+description: "Description of hackathon challenges."
+---
+
+## Challenges
+
+You've made it so far.
+Now the time has come to show the Unikraft world what you are really made of.
+A true Dragon Slayer!
+
```
Comment (razvand) [inline content/guides/challenges.mdx:11]:
```suggestion
We've prepared challenges that you will pass through to prove your worthiness.
Each challenge will get you points of valor, dubbed **HP** (_hackathon points_), that will be collected in the hackathon scoreboard.
```

File: content/guides/challenges.mdx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+See the items added in the [`Hackathons` project](https://github.com/orgs/unikraft/projects/29).
+
+You will get the number of points for submitting an issue or a solution that works.
+Those points will be doubled if the PR is accepted upstream.
```
Comment (razvand) [inline content/guides/challenges.mdx:15]:
```suggestion
Solutions are submitted as pull requests.
If a pull request is accepted upstream, you get double points.
```

File: content/guides/challenges.mdx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+After solving one challenge, pick another one.
+We recommend you start small then move up to more difficult challenges.
+
+Please check the community guidelines on [submitting changes](https://github.com/unikraft/docs/blob/2ff2de59b4032fb6ebefc05c24ef832adacf7768/content/en/docs/contributing/submitting-changes.md) and [the review process](https://github.com/unikraft/docs/blob/2ff2de59b4032fb6ebefc05c24ef832adacf7768/content/en/docs/contributing/review-process.md).
+
+### Challenge Types
+
+Challenge types generally fall into the categories below.
+You can go beyond the items marked in the [`Hackathons` project](https://github.com/orgs/unikraft/projects/29).
```
Comment (razvand) [inline content/guides/challenges.mdx:24]:
```suggestion
You can go beyond the items marked in the [`Hackathons` GitHub project](https://github.com/orgs/unikraft/projects/29).
```

File: content/guides/challenges.mdx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+1. **Submitting issues**
+
+   Test as many applications as you can, with different configuration options (e.g. different architectures, different filesystems, etc.).
```
Comment (razvand) [inline content/guides/challenges.mdx:28]:
```suggestion
   Test as many [Unikraft applications](https://github.com/search?q=topic%3Aunikraft-application+org%3Aunikraft&type=Repositories) as you can, with different configuration options (e.g. different architectures, different filesystems, etc.).
   Use both [KraftKit](https://github.com/unikraft/kraftkit) and the Make-based system for building, and KraftKit and `qemu-system` for running.
```

File: content/guides/challenges.mdx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+   Submit any issue you bump into.
+
+   5 hackathon points are awarded for each submitted issue.
```
Comment (razvand) [inline content/guides/challenges.mdx:31]:
```suggestion
   5 HP are awarded for each submitted issue.
```

File: content/guides/challenges.mdx [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+We recommend you start small then move up to more difficult challenges.
+
+Please check the community guidelines on [submitting changes](https://github.com/unikraft/docs/blob/2ff2de59b4032fb6ebefc05c24ef832adacf7768/content/en/docs/contributing/submitting-changes.md) and [the review process](https://github.com/unikraft/docs/blob/2ff2de59b4032fb6ebefc05c24ef832adacf7768/content/en/docs/contributing/review-process.md).
+
+### Challenge Types
+
+Challenge types generally fall into the categories below.
+You can go beyond the items marked in the [`Hackathons` project](https://github.com/orgs/unikraft/projects/29).
+
+1. **Submitting issues**
```
Comment (razvand) [inline content/guides/challenges.mdx:26]:
```suggestion
1. **Submit issues**
```

File: content/guides/challenges.mdx [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+   Test as many applications as you can, with different configuration options (e.g. different architectures, different filesystems, etc.).
+   Submit any issue you bump into.
+
+   5 hackathon points are awarded for each submitted issue.
+
+1. **Building applications as dynamic PIE binaries** and adding them to the [`dynamic-apps` repository](https://github.com/unikraft/dynamic-apps)
```
Comment (razvand) [inline content/guides/challenges.mdx:33]:
```suggestion
1. **Build or extract applications as dynamic PIE binaries** and add them to the [`dynamic-apps` repository](https://github.com/unikraft/dynamic-apps)
```

File: content/guides/challenges.mdx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+   These applications aren't required to run on Unikraft.
```
Comment (razvand) [inline content/guides/challenges.mdx:35]:
```suggestion

   You need to submit a directory, representing the root filesystem of the application, consisting of:
   
   - the application binary ELF
   - required application dynamic libraries (shared objects)
   - application-specific configuration, data or runtime files
   - a `README.md` detailing how the application is to be built, configured and run on Linux and on Unikraft;
     as an example, see the `README.md` for [Python](https://github.com/unikraft/dynamic-apps/tree/master/lang/python)

   These applications aren't required to run on Unikraft.
```

File: content/guides/challenges.mdx [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+   If something doesn't work submit an issue and get points for that.
```
Comment (razvand) [inline content/guides/challenges.mdx:36]:
```suggestion
   If something doesn't work submit an issue and get HP for that.
```

File: content/guides/challenges.mdx [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+   30-50 hackathon points are awarded for each submitted application, depending on its complexity.
```
Comment (razvand) [inline content/guides/challenges.mdx:38]:
```suggestion
   30-50 HP are awarded for each submitted application, depending on its complexity.
```

File: content/guides/challenges.mdx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+
+1. **Fixing build warning messages**
```
Comment (razvand) [inline content/guides/challenges.mdx:40]:
```suggestion
1. **Fix build warning messages**
```

File: content/guides/challenges.mdx [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+   5-10 hackathon points are awarded for each fix.
+
+1. **Updating library versions** (and making sure the new version builds and runs)
```
Comment (razvand) [inline content/guides/challenges.mdx:44]:
```suggestion
1. **Update library versions** (and making sure the new version builds and runs)

   Unikraft uses multiple [external libraries](https://github.com/search?q=topic%3Aunikraft-library+org%3Aunikraft&type=Repositories) that are required by [Unikraft applications](https://github.com/search?q=topic%3Aunikraft-application+org%3Aunikraft&type=Repositories) to run.
   These libraries are developed outside Unikraft by other open source communities.
   Periodically, we want to update the versions used by Unikraft, to benefit from most recent features.
   
   Steps are:
   
   1. Choose an [external library repository](https://github.com/search?q=topic%3Aunikraft-library+org%3Aunikraft&type=Repositories).
   1. Choose the upstream version you want to use, probably the most recent one.
   1. Update the build, configuration and source code files inside it to the  updated version.
   1. Test with all Unikraft applications using that library.
   1. Submit changes as a pull request in the library repository.
   1. Get HP.
```

File: content/guides/challenges.mdx [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+   20-30 hackathon points are usually awarded for the boost.
```
Comment (razvand) [inline content/guides/challenges.mdx:46]:
```suggestion
   20-30 HP are usually awarded for the boost.
```

File: content/guides/challenges.mdx [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+1. **Fixing issues**
```
Comment (razvand) [inline content/guides/challenges.mdx:48]:
```suggestion
1. **Fix issues**

   Look for issues in the [core `unikraft` repository](https://github.com/unikraft/unikraft/) or [any other repository in the `unikraft` GitHub organization](https://github.com/orgs/unikraft/repositories).
   Select an issue that intrigues you;
   it's good to first ask for feedback from the hackathon mentors, to ensure it's feasible to be solved during the hackathon.
   Proceed to work on the issue.
   Once solved, test it as much as possible.
   Submit solution as pull request and get HP.
```

File: content/guides/challenges.mdx [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+   Hackathon points will be awarded according to the complexity of the fix.
```
Comment (razvand) [inline content/guides/challenges.mdx:50]:
```suggestion
   HP will be awarded according to the complexity of the fix.
```

**Review by razvand — CHANGES_REQUESTED** [review]
See my suggestions.

**Comment by StefanJum** [comment]
Solved @razvand

File: content/guides/challenges.mdx [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+description: "Listing of hackathon challenges, together with awarded points"
+---
+
+## Hackathon Challenges
+
+You've been part of a Unikraft hackathon and you've made it so far.
+Now the time has come to show the Unikraft world what you are really made of.
+A true Dragon Slayer!
+
+We've prepared challenges that you will pass through to prove your worthiness.
```
Comment (razvand) [inline content/guides/challenges.mdx:12]:
```suggestion
We've prepared quite a number of challenges for you.
Prove your worthiness by passing them and, in doing so, adding your name to the list of hero contributors to the Unikraft project.
```

File: content/guides/challenges.mdx [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+Each challenge will get you points of valor, dubbed **HP** (_hackathon points_), that will be collected in the hackathon scoreboard.
```
Comment (razvand) [inline content/guides/challenges.mdx:13]:
```suggestion
Each challenge will get you points of valor, dubbed **HP** (_hackathon points_), that will be collected in the hackathon scoreboard of valor.
```

File: content/guides/challenges.mdx [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+See the items added in the [`Hackathons` GitHub project](https://github.com/orgs/unikraft/projects/29).
+
+You will get the number of points for submitting an issue or a solution that works.
+Solutions are submitted as pull requests.
+If a pull request is accepted upstream, you get double points.
```
Comment (razvand) [inline content/guides/challenges.mdx:19]:
```suggestion
If a pull request is accepted upstream, you get double HP.
```

File: content/guides/challenges.mdx [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+After solving one challenge, pick another one.
+We recommend you start small then move up to more difficult challenges.
```
Comment (razvand) [inline content/guides/challenges.mdx:21]:
```suggestion
We recommend you start small, then move up to more difficult challenges.
```

File: content/guides/challenges.mdx [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+
+Please check the community guidelines on [submitting changes](https://github.com/unikraft/docs/blob/2ff2de59b4032fb6ebefc05c24ef832adacf7768/content/en/docs/contributing/submitting-changes.md) and [the review process](https://github.com/unikraft/docs/blob/2ff2de59b4032fb6ebefc05c24ef832adacf7768/content/en/docs/contributing/review-process.md).
+
+### Challenge Types
+
+Challenge types generally fall into the categories below.
```
Comment (razvand) [inline content/guides/challenges.mdx:27]:
```suggestion
Challenges generally fall into the categories below.
```

File: content/guides/challenges.mdx [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+You can go beyond the items marked in the [`Hackathons` GitHub project](https://github.com/orgs/unikraft/projects/29).
```
Comment (razvand) [inline content/guides/challenges.mdx:28]:
```suggestion
You can venture beyond the challenges in the [`Hackathons` GitHub project](https://github.com/orgs/unikraft/projects/29).
```

File: content/guides/challenges.mdx [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+
+1. **Submit issues**
+
+   Test as many [Unikraft applications](https://github.com/search?q=topic%3Aunikraft-application+org%3Aunikraft&type=Repositories) as you can, with different configuration options (e.g. different architectures, different filesystems, etc.).
```
Comment (razvand) [inline content/guides/challenges.mdx:32]:
```suggestion
   Test as many [Unikraft applications](https://github.com/search?q=topic%3Aunikraft-application+org%3Aunikraft&type=Repositories) as you can, with different configuration options (i.e. different architectures, different filesystems, etc.).
```

File: content/guides/challenges.mdx [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+   Use both [KraftKit](https://github.com/unikraft/kraftkit) and the Make-based system for building, and KraftKit and `qemu-system` for running.
```
Comment (razvand) [inline content/guides/challenges.mdx:33]:
```suggestion
   Use both [KraftKit](https://github.com/unikraft/kraftkit) and the Make-based system for building, and boh KraftKit and `qemu-system` for running.
```

File: content/guides/challenges.mdx [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+
+   You need to submit a directory, representing the root filesystem of the application, consisting of:
+
+   - the application binary ELF
+   - required application dynamic libraries (shared objects)
+   - application-specific configuration, data or runtime files
+   - a `README.md` detailing how the application is to be built, configured and run on Linux and on Unikraft;
+    as an example, see the `README.md` for [Python](https://github.com/unikraft/dynamic-apps/tree/master/lang/python)
+
+   These applications aren't required to run on Unikraft.
```
Comment (razvand) [inline content/guides/challenges.mdx:48]:
```suggestion
   These applications aren't required to run on Unikraft in order to do a submission and get HP.
```

File: content/guides/challenges.mdx [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+   If something doesn't work submit an issue and get HP for that.
+
+   30-50 HP are awarded for each submitted application, depending on its complexity.
+
+1. **Fix build warning messages**
```
Comment (razvand) [inline content/guides/challenges.mdx:53]:
```suggestion
1. **Fix build warning messages**

    Use different build environments (different operating systems, different compilers), different targets (different platforms, different architectures) and different configuration options (different filesystems, different library support, debugging options) to configure and build [Unikraft applications](https://github.com/search?q=topic%3Aunikraft-application+org%3Aunikraft&type=Repositories).
    If there are build warnings, create a fix and submit a pull request.
```

File: content/guides/challenges.mdx [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+
+   5-10 HP are awarded for each fix.
+
+1. **Update library versions** (and making sure the new version builds and runs)
```
Comment (razvand) [inline content/guides/challenges.mdx:57]:
```suggestion
1. **Update library versions** (and make sure the new version builds and runs)
```

File: content/guides/challenges.mdx [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+
+   Unikraft uses multiple [external libraries](https://github.com/search?q=topic%3Aunikraft-library+org%3Aunikraft&type=Repositories) that are required by [Unikraft applications](https://github.com/search?q=topic%3Aunikraft-application+org%3Aunikraft&type=Repositories) to run.
+   These libraries are developed outside Unikraft by other open source communities.
+   Periodically, we want to update the versions used by Unikraft, to benefit from most recent features.
+
+   Steps are:
+
+   1. Choose an [external library repository](https://github.com/search?q=topic%3Aunikraft-library+org%3Aunikraft&type=Repositories).
+   1. Choose the upstream version you want to use, probably the most recent one.
+   1. Update the build, configuration and source code files inside it to the  updated version.
```
Comment (razvand) [inline content/guides/challenges.mdx:67]:
```suggestion
   1. Update the build, configuration and source code files inside it to the updated version.
```

File: content/guides/challenges.mdx [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+   1. Test with all Unikraft applications using that library.
+   1. Submit changes as a pull request in the library repository.
+   1. Get HP.
+
+   20-30 HP are usually awarded for the boost.
```
Comment (razvand) [inline content/guides/challenges.mdx:72]:
```suggestion
   20-30 HP are usually awarded for each library update.
```

File: content/guides/challenges.mdx [hunk 34]
```diff
@@ -0,0 +1,83 @@
...
+
+1. **Fixing issues**
```
Comment (razvand) [inline content/guides/challenges.mdx:74]:
```suggestion
1. **Fix issues**
```

File: content/guides/challenges.mdx [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+
+   Look for issues in the [core `unikraft` repository](https://github.com/unikraft/unikraft/) or [any other repository in the `unikraft` GitHub organization](https://github.com/orgs/unikraft/repositories).
+   Select an issue that intrigues you;
+   it's good to first ask for feedback from the hackathon mentors, to ensure it's feasible to be solved during the hackathon.
```
Comment (razvand) [inline content/guides/challenges.mdx:78]:
```suggestion
   it's good to first ask for feedback from the hackathon mentors, to ensure it's feasible to solve the issue during the hackathon.
```

File: content/guides/challenges.mdx [hunk 36]
```diff
@@ -0,0 +1,83 @@
...
+   Proceed to work on the issue.
+   Once solved, test it as much as possible.
+   Submit solution as pull request and get HP.
```
Comment (razvand) [inline content/guides/challenges.mdx:81]:
```suggestion
   Submit solution as a pull request and get HP.
```

**Review by razvand — CHANGES_REQUESTED** [review]
round 2

**Comment by StefanJum** [comment]
Done @razvand

File: content/guides/challenges.mdx [hunk 37]
```diff
@@ -0,0 +1,87 @@
...
+   You need to submit a directory, representing the root filesystem of the application, consisting of:
+
+   - the application binary ELF
+   - required application dynamic libraries (shared objects)
+   - application-specific configuration, data or runtime files
+   - a `README.md` detailing how the application is to be built, configured and run on Linux and on Unikraft;
+    as an example, see the `README.md` for [Python](https://github.com/unikraft/dynamic-apps/tree/master/lang/python)
+
+   These applications aren't required to run on Unikraft in order to do a submission and get HP.
+   If something doesn't work submit an issue and get HP for that.
```
Comment (razvand) [inline content/guides/challenges.mdx:50]:
```suggestion
   If something doesn't work, submit an issue and get HP for that.
```

**Review by razvand — APPROVED** [review]
Thanks, @StefanJum. Looks good.

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**